### ISSUE TITLE
Exception thrown if single aspect type has multiple constructor overloads of same length

### ISSUE THREAD

**Issue by keith-anders** [vescon/MethodBoundaryAspect.Fody#23]
Create an aspect that has multiple constructor overloads of the same length. Use one of those overloads, and an exception will be thrown during weaving.

I have a fix ready for it that I could submit a PR for, but there's no "contributing" file, so I don't know whether you would accept a PR from just anyone or what your standards are.

If you want to see what I did so you can review it, [here is the branch](https://github.com/keith-anders/MethodBoundaryAspect.Fody/tree/fix-attribute-parameter-issue).

My PR also fixes type arguments not being supported in the aspect ctor, array types not being supported in the aspect ctor, and named field assignments in aspect creation.

**Comment by maulwuff** [comment]
this ticket was added in PR #24 and is currently live in 1.0.16 but breaks existing code - in my case.
```
6>    Fody: Fody (version 3.0.3.0) Executing
6>MSBUILD : error : Fody: An unhandled exception occurred:
6>MSBUILD : error : Exception:
6>MSBUILD : error : Failed to execute weaver ...\packages\MethodBoundaryAspect.Fody.1.0.16\MethodBoundaryAspect.Fody.dll
6>MSBUILD : error : Type:
6>MSBUILD : error : System.Exception
6>MSBUILD : error : StackTrace:
6>MSBUILD : error :    at InnerWeaver.ExecuteWeavers() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 208
6>MSBUILD : error :    at InnerWeaver.Execute() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 103
6>MSBUILD : error : Source:
6>MSBUILD : error : FodyIsolated
6>MSBUILD : error : TargetSite:
6>MSBUILD : error : Void ExecuteWeavers()
6>MSBUILD : error : Didn't found matching constructor on type 'Tsi.Reusable.Generics.CodeWeaving.TriggerPropChangedAttribute'
6>MSBUILD : error : Type:
6>MSBUILD : error : System.InvalidOperationException
6>MSBUILD : error : StackTrace:
6>MSBUILD : error :    at MethodBoundaryAspect.Fody.InstructionBlockCreator.NewObject(VariableDefinition newInstance, TypeReference instanceTypeReference, ModuleDefinition module, CustomAttribute aspect)
6>MSBUILD : error :    at MethodBoundaryAspect.Fody.InstructionBlockChainCreator.CreateAspectInstance(CustomAttribute aspect)
6>MSBUILD : error :    at MethodBoundaryAspect.Fody.MethodWeaver.Weave(ModuleDefinition module, MethodDefinition method, IEnumerable`1 aspectInfos)
```

I narrowed it down to this code to create the exception above:
as soon as I call a ctor with a single parameter from the decorated property, it fails.
(it also fails with 2 parameters ;) )

```csharp
public class TestClass
{
        [TriggerPropChanged("a")]
        public string TestProp { get;set; }
 }

public class TriggerPropChangedAttribute : OnMethodBoundaryAspect
{
   public TriggerPropChangedAttribute(string alsoNotify )  //this is the only c'tor!
    {  /* ... */   }
}
```

**Comment by keith-anders** [comment]
The following compiles without issue using 1.0.16 in either .net Framework or .net Standard.

```csharp
public class TestClass
{
  [TriggerPropChanged("a")]
  public string TestProp { get; set; }
}

public class TriggerPropChangedAttribute : OnMethodBoundaryAspect
{
  public TriggerPropChangedAttribute(string alsoNotify) { }

  public override void OnEntry(MethodExecutionArgs arg) => base.OnEntry(arg);
  public override void OnException(MethodExecutionArgs arg) => base.OnException(arg);
  public override void OnExit(MethodExecutionArgs arg) => base.OnExit(arg);
}```

@maulwuff  Can you help clarify what I am missing in order to set up a reproducible example of your issue?

**Comment by maulwuff** [comment]
The key is, to have the attribute in another dll than the decorated class. 
Find a sample project attached. It contains two projects, both of them have a class to weave.

- project `MethodBoundaryTest` will compile and weave just fine (it holds the attribute implementation)
- project `UnitTestProject` will throw `Didn't found matching constructor`

[MethodBoundaryTest.zip](https://github.com/vescon/MethodBoundaryAspect.Fody/files/2003846/MethodBoundaryTest.zip)

expected result: 
- fails to build with 1.0.16
- succeeds with 1.0.12  (make sure to clean packages folder and `C:\Users\yourname\.nuget\packages\methodboundaryaspect.fody\` from any newer version than 1.0.12!!! It will pull in the newer version if there is any)


dang, where did 1.0.23 just came from?

**Comment by Ralf1108** [comment]
@keith-anders is this issue then resolved with your PR #27?

**Comment by keith-anders** [comment]
My issue is resolved on 1.0.41. Thanks.

**Comment by maulwuff** [comment]
works for me, too :+1:
